numpy/core/src/multiarray/item_selection.c

```diff
@@ -2096,10 +2103,13 @@ PyArray_CountNonzero(PyArrayObject *self)
             if (nonzero(data, self)) {
                 ++nonzero_count;
             }
+            if (needs_api && PyErr_Occurred()) {
+                return -1;
+            }
             data += stride;
         }
 
-        return nonzero_count;
+        return PyErr_Occurred() ? -1 : nonzero_count;
```

Comment:
It seems plausible to me that nonzero returns a special value on error, might want to use that as well. Ufunc machinery checks for error in the outer loop, the inner loop function should return on error though.